clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java

```diff
@@ -0,0 +1,221 @@
+     * reconnect backoff window following the disconnect).
+     */
+    public boolean nodeUnavailable(Node node) {
+        return client.connectionFailed(node) && client.connectionDelay(node, time.milliseconds()) > 0;
+    }
+
+    public void tryDisconnect(Optional<Node> coordinator) {
+        coordinator.ifPresent(node -> client.disconnect(node.idString()));
+    }
+
+    public void close() throws IOException {
+        this.client.close();
+    }
+
+    public static class UnsentRequest {
+        private final Optional<RequestFutureCompletionHandlerBase> callback;
+        private final AbstractRequest.Builder abstractBuilder;
+        private final Optional<Node> node; // empty if random node can be choosen
+        private final Timer timer;
+
+        public UnsentRequest(final Timer timer,
+                             final AbstractRequest.Builder abstractBuilder,
+                             final RequestFutureCompletionHandlerBase callback) {
+            this(timer, abstractBuilder, callback, null);
+        }
+
+        public UnsentRequest(final Timer timer,
+                             final AbstractRequest.Builder abstractBuilder,
+                             final RequestFutureCompletionHandlerBase callback,
+                             final Node node) {
+            Objects.requireNonNull(abstractBuilder);
+            this.abstractBuilder = abstractBuilder;
+            this.node = Optional.ofNullable(node);
+            this.callback = Optional.ofNullable(callback);
+            this.timer = timer;
+        }
+    }
+
+    public static class RequestFutureCompletionHandlerBase implements RequestCompletionHandler {
+        private final RequestFuture<ClientResponse> future;
+        private ClientResponse response;
+        private RuntimeException e;
+
+        RequestFutureCompletionHandlerBase() {
+            this.future = new RequestFuture<>();
+        }
+
+        public void fireCompletion() {
+            if (e != null) {
+                future.raise(e);
+            } else if (response.authenticationException() != null) {
+                future.raise(response.authenticationException());
+            } else if (response.wasDisconnected()) {
+                //log.debug("Cancelled request with header {} due to node {} being disconnected", response
+                // .requestHeader(), response.destination());
+                future.raise(DisconnectException.INSTANCE);
+            } else if (response.versionMismatch() != null) {
+                future.raise(response.versionMismatch());
+            } else {
+                future.complete(response);
+            }
+        }
+
+        public void onFailure(RuntimeException e) {
+            this.e = e;
+            fireCompletion();
+            handleResponse(response, e);
+        }
+
+        public void handleResponse(ClientResponse r, Throwable t) {}
```

Comment:
I thought about it, but I think for the case of fetch request we actually don't necessary need to `handleResponse` so I thought it would be easier to make it a base class, and only extend it when needed.

Reply:
Why would we not need it for fetch requests?